components/vfs/include/components/vfs/filesystem.hpp

```diff
@@ -121,14 +121,15 @@ class FileSystem
 	virtual size_t        file_size(const std::string &file_path)                                                                        = 0;
 	virtual StackErrorPtr write_file(const std::string &file_path, const void *data, size_t size)                                        = 0;
 	virtual void          make_directory(const std::string &path)                                                                        = 0;
+	virtual bool          remove(const std::string &path)                                                                                = 0;
 
 	virtual StackErrorPtr enumerate_files(const std::string &file_path, std::vector<std::string> *files)     = 0;
```

Comment:
Why are you using `std::vector<std::string> *`, instead of `std::vector<std::string> &` ?
The question holds for all the other functions, where you pass in a pointer instead of a reference, as well.

Reply:
I have always preferred the C style output parameters over pass by reference. From the usage it is more obvious which elements are being mutated.

`enumarate_files(path, &files)` compared to `enumarate_files(path, files)`

Here it is clear to see that `files` will be mutated in some way. This is mainly due to the fact I have had strict requirements on older C++ versions where variant was not yet part of the std lib. If we want to use more variants and have Go style returns of multiple variables then I would be happy to change the API.

Will `std::variant<StackErrorPtr, std::vector<std::string>>` incur more copies?

Reply:
It's a fairly common idiom to pass const params by reference and mutable params by pointer. As you say, it's obvious from the call-site that a parameter may be changed by the call. I don't personally see any issue with it since it's what I also do.

Reply:
When you pass a mutable parameter by pointer, you have to check against nullptr. When you pass it by reference you don't need to.
Either way, you don't know if the passed parameter is reset, or just modified (here, it could as well be that the enumerated files are appended to the vector of files); when you return something, it's clearly new.

Reply:
So we would be going form this style:

```c++
Thing something;

if (auto err = action(&something)) {
    // we hit an error
    fix_this(err);
}

// no error
something.do();
```

to this style:

```c++
auto result = action(); // returns std::variant<StackErrorPtr, Thing>

if (auto err = std::get<StackErrorPtr>(result)) {
    // we hit an error
    fix_this(err);
}

// no error
Thing  something = std::get<Thing>(result);
```

in practice the change would look more like this:

```c++
template <typename Type>
using StackErrorVariant = std::variant<StackErrorPtr, Type>

virtual StackErrorVariant<std::vector<std::string>> enumerate_files(const std::string &file_path) = 0; 
```

Thoughts @SaschaWillems @gpx1000 @gary-sweet ?

Reply:
Would switching from those StackErrorPtr to exceptions be an option here? Would simplify the function even more:
`std::vector<std::string> enumerate_files( const std::string &file_path) = 0;`

Reply:
It might, but exceptions just move the verbosity up a level. For now this interface works and so it may be best to keep it until there are major hurdles caused by it. At the end of the day this is all just syntactic sugar and none of the options would not change the behavior of the system.

Any further opinions on this @gpx1000 @SaschaWillems @gary-sweet? It would make sense to find a common ground on error handling sooner rather than later, but should be addressed in a new PR

Reply:
I personally prefer pointers to references.  It's easier for me to see who actually originates the memory in a passing context and where it must be created/deleted.  However, that's just easier for older guys like me that remember the C days.  The reference method and new tooling makes this more of a spaces/tabs argument.  Both work equally well and if anything, points are in favor of pass by reference due to the fact that pointers are way easier to have security problems and get lost causing memory leaks etc.

As for the exceptions instead of StackErrorPtr, I'm a fan of extensions.  I like programs to break when they hit real problems that put them into a non-working state.  Break early break often is usually the path towards highlighting errors sooner in the dev process.  That said, I agree with Tom that in this instance, I don't see it providing much new functionality.  It would add to the size of the executable and is about 16 times slower to switch to them for no appreciable benefit (not that anyone really cares about exe size anymore, and why care about CPU performance when you want to throw an error).  This one I'm more ambivalent about.

Reply:
> but exceptions just move the verbosity up a level

I don't really understand this. What do you mean?

Reply:
The function definition looks cleaner when using exceptions but handling the error is more verbose with try/catch compared to the `if` style. It is also not clear by the function definition that it throws an error. We can get lazy and have multiple functions which throw inside the same try/catch statement, it might be recoverable earlier in the process.

Reply:
> It is also not clear by the function definition that it throws an error

That's not exactly true... as this function uses a non-const std::vector, it might throw anyways, no matter how other errors might be communicated.

Reply:
Ok, so we have hit a little bit of a stale mate. Both techniques have pros and cons but essentially do the same thing. As it is not the purpose of this PR to address error handling, are you happy to approve this PR as is?

We should either raise a ticket which moves this discussion there or you can create a proposal PR for how we should handle errors going forward. I'm not opposed to exceptions, simply using the method that I have found works best for me in the past. This is not the PR that should address it :)

The only real requirement in terms of errors is that we have a consistent error handling mechanism across the code base so that errors are predictable :)

Reply:
> The only real requirement in terms of errors is that we have a consistent error handling mechanism across the code base so that errors are predictable :)

Totally agree!